src/ImageSharp/Memory/Allocators/ArrayPoolMemoryAllocator.cs

```diff
@@ -12,140 +13,131 @@ namespace SixLabors.ImageSharp.Memory
-            this.maxArraysPerBucketNormalPool = maxArraysPerBucketNormalPool;
-
-            this.InitArrayPools();
+            this.MaxPooledArrayLengthInBytes = maxPooledArrayLengthInBytes;
+            this.MaxContiguousArrayLengthInBytes = maxContiguousArrayLengthInBytes;
+            this.MaxArraysPerPoolBucket = maxArraysPerPoolBucket;
+            this.largeArrayPool = new GCAwareConfigurableArrayPool<byte>(this.MaxPooledArrayLengthInBytes, this.MaxArraysPerPoolBucket);
         }
 
         /// <summary>
-        /// Gets the maximum size of pooled arrays in bytes.
+        /// Gets the maximum length, in bytes, of an array instance that may be stored in the pool.
         /// </summary>
-        public int MaxPoolSizeInBytes { get; }
+        public int MaxPooledArrayLengthInBytes { get; }
 
         /// <summary>
-        /// Gets the threshold to pool arrays in <see cref="largeArrayPool"/> which has less buckets for memory safety.
+        /// Gets the maximum number of array instances that may be stored in each bucket in the pool.
         /// </summary>
-        public int PoolSelectorThresholdInBytes { get; }
+        public int MaxArraysPerPoolBucket { get; }
 
         /// <summary>
-        /// Gets the length of the largest contiguous buffer that can be handled by this allocator instance.
+        /// Gets the maximum length of the largest contiguous buffer that can be handled by this allocator instance.
         /// </summary>
-        public int BufferCapacityInBytes { get; internal set; } // Setter is internal for easy configuration in tests
+        public int MaxContiguousArrayLengthInBytes { get; internal set; } // Setter is internal for easy configuration in tests
 
         /// <inheritdoc />
         public override void ReleaseRetainedResources()
-        {
-            this.InitArrayPools();
-        }
+            => this.largeArrayPool.Trim();
 
         /// <inheritdoc />
-        protected internal override int GetBufferCapacityInBytes() => this.BufferCapacityInBytes;
+        protected internal override int GetMaxContiguousArrayLengthInBytes() => this.MaxContiguousArrayLengthInBytes;
 
         /// <inheritdoc />
         public override IMemoryOwner<T> Allocate<T>(int length, AllocationOptions options = AllocationOptions.None)
         {
             Guard.MustBeGreaterThanOrEqualTo(length, 0, nameof(length));
+
             int itemSizeBytes = Unsafe.SizeOf<T>();
             int bufferSizeInBytes = length * itemSizeBytes;
-            if (bufferSizeInBytes < 0 || bufferSizeInBytes > this.BufferCapacityInBytes)
+
+            IMemoryOwner<T> memory;
+            if (bufferSizeInBytes > this.MaxPooledArrayLengthInBytes)
             {
-                ThrowInvalidAllocationException<T>(length);
+                // For anything greater than our pool limit defer to unmanaged memory
+                // to prevent LOH fragmentation.
+                memory = new UnmanagedBuffer<T>(length);
```

Comment:
Would be great to see some metrics, how many times do we go here VS to the pool.
I can help with the EventSource stuff.

Reply:
I know the BMP codec hits this a few times. I saw that when I made a mistake in the `UnmanagedBuffer<T>` type. Hopefully my EventSource stuff works ok.